File: javascript/socketd/src/socketd/transport/core/Session.ts
import type {Entity, Reply} from "./Entity";
import type {Message} from "./Message";
import type {IoConsumer} from "./Typealias";
import type {Channel} from "./Channel";
import type {Stream} from "./Stream";
import type {ClientSession} from "../client/ClientSession";
import type {Handshake} from "./Handshake";


/**
 * 会话
 *
 * @author noear
 * @since 2.0
 */
export interface Session extends ClientSession {
    /**
     * 获取握手信息
     */
    handshake(): Handshake;

    /**
     * broker player name
     *
     * @since 2.1
     */
    name(): string | undefined;

    /**
     * 获取握手参数
     *
     * @param name 名字
     */
    param(name: string): string | undefined;

    /**
     * 获取握手参数或默认值
     *
     * @param name 名字
     * @param def  默认值
     */
    paramOrDefault(name: string, def: string): string;

    /**
     * 获取握手路径
     */
    path(): string;

    /**
     * 设置握手新路径
     */
    pathNew(pathNew: string);

    /**
     * 获取所有属性
     */
    attrMap(): Map<string, any>;

    /**
     * 是有属性
     *
     * @param name 名字
     */
    attrHas(name: string);

    /**
     * 获取属性
     *
     * @param name 名字
     */
    attr(name: string): any;

    /**
     * 获取属性或默认值
     *
     * @param name 名字
     * @param def  默认值
     */
    attrOrDefault(name: string, def: object): object;

    /**
     * 设置属性
     *
     * @param name  名字
     * @param val 值
     */
    attrPut(name: string, val: object);

    /**
     * 手动发送 Ping（一般是自动）
     */
    sendPing();

    /**
     * 发送告警
     */
    sendAlarm(from: Message, alarm: string);

    /**
     * 答复
     *
     * @param from    来源消息
     * @param content 内容
     */
    reply(from: Message, content: Entity);

    /**
     * 答复并结束（即最后一次答复）
     *
     * @param from    来源消息
     * @param content 内容
     */
    replyEnd(from: Message, content: Entity);
}

/**
 * 会话基类
 *
 * @author noear
 */
export abstract class SessionBase implements Session {
    protected _channel: Channel;
    private _sessionId: string;
    private _attrMap: Map<string, object>;

    constructor(channel: Channel) {
        this._channel = channel;
        this._sessionId = this.generateId();
    }

    sessionId(): string {
        return this._sessionId;
    }

    name(): string | undefined{
        return this.param("@");
    }

    attrMap(): Map<string, any> {
        if (this._attrMap == null) {
            this._attrMap = new Map<string, any>();
        }

        return this._attrMap;
    }

    attrHas(name: string) {
        if (this._attrMap == null) {
            return false;
        }

        return this._attrMap.has(name);
    }

    attr(name: string): any {
        if (this._attrMap == null) {
            return null;
        }

        return this._attrMap.get(name);
    }

    attrOrDefault(name: string, def: object): object {
        const tmp = this.attr(name);
        return tmp ? tmp : def;
    }

    attrPut(name: string, val: object) {
        this.attrMap().set(name, val);
    }

    abstract handshake(): Handshake ;

    abstract param(name: string): string | undefined;

    abstract paramOrDefault(name: string, def: string): string;

    abstract path(): string ;

    abstract pathNew(pathNew: string);

    abstract sendPing();

    abstract sendAlarm(from: Message, alarm: string);

    abstract reply(from: Message, entity: Entity);

    abstract replyEnd(from: Message, entity: Entity);

    abstract isValid(): boolean ;

    abstract reconnect();

    abstract send(event: string, content: Entity);

    abstract sendAndRequest(event: string, content: Entity, callback: IoConsumer<Reply>, timeout?: number): Stream;

    abstract sendAndSubscribe(event: string, content: Entity, callback: IoConsumer<Reply>, timeout?: number): Stream;

    abstract close();

    protected generateId() {
        return this._channel.getConfig().getIdGenerator().generate();
    }
}

File: java/socketd/src/main/java/org/noear/socketd/transport/core/Session.java
package org.noear.socketd.transport.core;

import org.noear.socketd.transport.client.ClientSession;
import org.noear.socketd.utils.IoConsumer;

import java.io.Closeable;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.util.Map;

/**
 * 会话
 *
 * @author noear
 * @since 2.0
 */
public interface Session extends ClientSession, Closeable {
    /**
     * 获取远程地址
     */
    InetSocketAddress remoteAddress() throws IOException;

    /**
     * 获取本地地址
     */
    InetSocketAddress localAddress() throws IOException;

    /**
     * 获取握手信息
     */
    Handshake handshake();

    /**
     * broker player name
     *
     * @since 2.1
     */
    default String name() {
        return param("@");
    }

    /**
     * 获取握手参数
     *
     * @param name 名字
     */
    String param(String name);

    /**
     * 获取握手参数或默认值
     *
     * @param name 名字
     * @param def  默认值
     */
    String paramOrDefault(String name, String def);

    /**
     * 获取握手路径
     */
    String path();

    /**
     * 设置握手新路径
     */
    void pathNew(String pathNew);

    /**
     * 获取所有属性
     */
    Map<String, Object> attrMap();

    /**
     * 是有属性
     *
     * @param name 名字
     */
    boolean attrHas(String name);

    /**
     * 获取属性
     *
     * @param name 名字
     */
    <T> T attr(String name);

    /**
     * 获取属性或默认值
     *
     * @param name 名字
     * @param def  默认值
     */
    <T> T attrOrDefault(String name, T def);

    /**
     * 放置属性
     *
     * @param name  名字
     * @param value 值
     */
    <T> Session attrPut(String name, T value);

    /**
     * 是否有效
     */
    boolean isValid();

    /**
     * 获取会话Id
     */
    String sessionId();

    /**
     * 手动重连（一般是自动）
     */
    void reconnect() throws IOException;

    /**
     * 手动发送 Ping（一般是自动）
     */
    void sendPing() throws IOException;

    /**
     * 发送告警
     */
    void sendAlarm(Message from, String alarm) throws IOException;

    /**
     * 发送
     *
     * @param event   事件
     * @param content 内容
     */
    void send(String event, Entity content) throws IOException;

    /**
     * 发送并请求
     *
     * @param event   事件
     * @param content 内容
     */
    default Reply sendAndRequest(String event, Entity content) throws IOException{
        return sendAndRequest(event, content, 0);
    }

    /**
     * 发送并请求（限为一次答复；指定超时）
     *
     * @param event   事件
     * @param content 内容
     * @param timeout 超时（毫秒）
     */
    Reply sendAndRequest(String event, Entity content, long timeout) throws IOException;

    /**
     * 发送并请求（限为一次答复；指定回调）
     *
     * @param event    事件
     * @param content  内容
     * @param consumer 回调消费者
     * @return 流
     */
    default Stream sendAndRequest(String event, Entity content, IoConsumer<Reply> consumer) throws IOException {
        return sendAndRequest(event, content, consumer, 0);
    }

    /**
     * 发送并请求（限为一次答复；指定回调）
     *
     * @param event    事件
     * @param content  内容
     * @param consumer 回调消费者
     * @param timeout  超时（毫秒）
     * @return 流
     */
    Stream sendAndRequest(String event, Entity content, IoConsumer<Reply> consumer, long timeout) throws IOException;

    /**
     * 发送并订阅（答复结束之前，不限答复次数）
     *
     * @param event    事件
     * @param content  内容
     * @param consumer 回调消费者
     * @return 流
     */
    default Stream sendAndSubscribe(String event, Entity content, IoConsumer<Reply> consumer) throws IOException {
        return sendAndSubscribe(event, content, consumer, 0);
    }

    /**
     * 发送并订阅（答复结束之前，不限答复次数）
     *
     * @param event    事件
     * @param content  内容
     * @param consumer 回调消费者
     * @param timeout  超时（毫秒）
     * @return 流
     */
    Stream sendAndSubscribe(String event, Entity content, IoConsumer<Reply> consumer, long timeout) throws IOException;

    /**
     * 答复
     *
     * @param from    来源消息
     * @param content 内容
     */
    void reply(Message from, Entity content) throws IOException;

    /**
     * 答复并结束（即最后一次答复）
     *
     * @param from    来源消息
     * @param content 内容
     */
    void replyEnd(Message from, Entity content) throws IOException;
}